src/lsm/scan_tree.zig

```diff
@@ -0,0 +1,666 @@
+        iterator: LevelTableValueBlockIterator,
+        buffer: LevelBuffer,
+
+        /// State over the manifest.
+        manifest: union(enum) {
+            begin,
+            next_table: ScanTree.Key,
+            finished,
+        },
+
+        /// State over the values.
+        values: union(enum) {
+            fetching,
+            buffered: RangeCursor,
+            finished,
+        },
+
+        tracer_slot: ?tracer.SpanStart = null,
+
+        pub inline fn init(scan: *ScanTree, buffer: LevelBuffer, level_index: u8) ScanTreeLevel {
+            return .{
+                .level_index = level_index,
+                .scan = scan,
+                .iterator = LevelTableValueBlockIterator.init(),
+                .buffer = buffer,
+                .manifest = .begin,
+                .values = .fetching,
+            };
+        }
+
+        /// Moves the level iterator to the next `table_info` that might contain the key range.
+        pub fn next_table(level: *ScanTreeLevel) void {
+            const scan: *ScanTree = level.scan;
+
+            const table_info: ?*const TableInfo = blk: {
+                const manifest: *Manifest = &scan.tree.manifest;
+                var key_exclusive: ?Key = switch (level.manifest) {
+                    .begin => null,
+                    .next_table => |key| key,
+                    .finished => unreachable,
+                };
+
+                if (manifest.next_table(.{
+                    .level = level.level_index,
+                    .snapshot = scan.snapshot,
+                    .key_min = scan.key_min,
+                    .key_max = scan.key_max,
+                    .key_exclusive = key_exclusive,
+                    .direction = scan.direction,
+                })) |next_table_info| {
+                    key_exclusive = switch (scan.direction) {
+                        .ascending => next_table_info.key_max,
+                        .descending => next_table_info.key_min,
+                    };
+
+                    if (key_exclusive.? < scan.key_min or
+                        key_exclusive.? > scan.key_max)
+                    {
+                        level.manifest = .finished;
+                    } else {
+                        level.manifest = .{ .next_table = key_exclusive.? };
+                    }
+                    break :blk next_table_info;
+                } else {
+                    level.manifest = .finished;
+                    break :blk null;
+                }
+            };
+
+            level.iterator.start(.{
+                .grid = scan.tree.grid,
+                .level = level.level_index,
+                .snapshot = scan.snapshot,
+                .index_block = level.buffer.index_block,
+                .tables = .{ .scan = table_info },
+                .direction = scan.direction,
+            });
+        }
+
+        pub fn fetch(level: *ScanTreeLevel) void {
+            assert(level.values == .fetching);
```

Comment:
In fact, we _can_ be already `.fetching` if `data_block_callback` returns no data, and we move to the next table.
I found it easier to make `.fetching` the initial state, instead of `.idle`.